docs: rollup bsn contract guide (#83)
babylonlabs-io/rollup-bsn-contracts: docs/deployment-integration.md

     2. Add text describing why someone would modify those.
        The document should provide understanding on how to be a
        contract admin, not just on what are the tech capabilities.
     3. We can also link to the documentations we have (e.g. `PRUNING.md`) -->

### Modifying the Contract Administrator

- `UpdateAdmin`: Transfers contract admin rights to a new Babylon address. Validates the address format and enforces 
access control using `cw_controllers::Admin`. Executed via `wasm execute` with the `UpdateAdmin` message

### Data Pruning

- `PruneData`: Removes all finality signatures and public randomness for rollup blocks with `height ≤ block_height`. 
Executed via `wasm execute` with the `PruneData` message. Optional parameters control 
batch size to prevent gas exhaustion

### Modifying the Finality Providers Allow-List
- `SetFPAllowlist`: Manages the FP allowlist by adding or removing FPs authorized to submit signatures and randomness

### Rate Limiting Submissions 
- `SetSubmissionInterval`: Updates the submission interval configuration to control how frequently each FP can 
submit signatures (e.g., every X-th block)

> <!-- TODO: what about public randomness parameters? -->

Author: What public randomness parameters? Currently we only have
`add_to_allowlist`, `remove_from_allowlist` , `update_admin`, `prune_data`

And non admin once
`submit_finality_signature`, `commit_public_randomness`

  "min_pub_rand": $MIN_PUB_RAND,
  "rate_limiting_interval": $RATE_LIMITING_INTERVAL,
  "allowed_finality_providers": ["$FP_PUBKEY_1", "$FP_PUBKEY_2"]
}
EOF
)

babylond tx wasm instantiate $CODE_ID '$INSTANT_MSG'
```

The parameters:
* `admin`: The Babylon Genesis address of the contract administrator
* `bsn_id`: The unique identifier for the BSN rollup
  Ensure that this is unique and not already used by another BSN on Babylon
  Genesis as this will affect your registration
* `min_pub_rand`: Minimum public randomness values included per public
  randomness commit by finality providers
* `rate_limiting_interval`: Length (in blocks) of the interval for
  rate-limiting submissions
* `max_msgs_per_interval`: Maximum number of submissions allowed per finality
  provider within each interval
* `allowed_finality_providers`: List of finality providers from which the contract will
  accept finality signatures and randomness

<!-- TODO: once we have docs for the allow list and rate limiting, we can link
those here -->

## Rollup BSN Registration
<img width="3032" height="287" alt="register" src="./assets/register.png" />

After deploying and instantiating the Rollup BSN contract, the

> rollup must register on Babylon Genesis. Registration requires 

Reviewer: register -> register as a BSN

submitting metadata to identify and describe the BSN

To register the BSN, use:
```shell
babylond tx btcstkconsumer register-consumer \
                                <consumer_id> \
                                <consumer_name> \
                                <consumer_ description> \
                                <max_multi_staked_fps> \
                                <rollup_finality_contract_address>
```

Required metadata for BSN registration:
* `consumer_id`: BSN ID of your rollup (same as the BSN ID you used on contract
  instantiation)
* `consumer_name`: Human-readable name of your rollup BSN (e.g., `"DeFi Rollup Chain"`)
* `consumer_description`: Brief description of the rollup BSN's purpose
* `max_multi_staked_fps`: Maximum number of rollup FPs per BTC delegation. Prevents delegators from splitting

>  their BTC across too many networks (recommended: `3-5`)

Reviewer: I think this feature has been removed from Babylon. Now Babylon enforces a global `max_multi_staked_fps` for all BSNs https://github.com/babylonlabs-io/babylon/blob/main/proto/babylon/btcstaking/v1/params.proto#L71-L73

choose for the deployment of the rollup BSN, the network might be permissioned:
* **Permissioned CosmWasm**: Requires governane approval for deploying a smart
  contract. This can be granted in two ways:
  * `MsgStoreCode` proposal: Upload a contract code. Preferred for one-time deployments
  * `MsgUpdateParams` proposal: Add a Babylon Genesis address in an allow-list
    for which its members can permissionlessly upload code. This is typically
    used for contracts requiring periodic maintenance.
  > To learn more about permissioned CosmWasm,
  > see the [deployment
  > guide](https://docs.babylonlabs.io/guides/governance/submit_proposals/smart_contract_deployment/) 
  > and review past proposals in the [governance
  > forum](https://forum.babylon.foundation/c/gov-proposals/smart-contract-proposals/15).
* **Permissioned BSN Registration**: Requires a governance approval for
  the registration of a BSN. This can be granted via the submission of
  a governance proposal executing a `MsgRegisterConsumer` using the
  metadata defined in the [Rollup BSN Registration](#rollup-bsn-registration)
  section.

  For governance guidelines, please refer to the governance forum: 

>   [Babylon Governance Guidelines](https://forum.babylon.foundation/c/gov-proposals/9/none)

Reviewer 2: Instead of having this text here, how about we add a TODO to fill this when there's a section for BSN registration props?

> **Note**: If a network uses both Permissioned CosmWasm and 
> Permissioned BSN Registration, two separate governance proposals 

> are currently required. A unified

Reviewer 2: ```suggestion
> are currently required. A unified
```

> governance flow requiring only a single governance proposal is being
> examined.

For the rest of the document, we will assume a Permissionless CosmWasm and BSN
Registration network for simplicity. Listings that would require governance
will be highlighted appropriately.

## Instantiation
<img width="3033" height="275" alt="instantiate" src="https://github.com/user-attachments/assets/f0c6f466-a3e9-475b-a256-d79ebcfc2033" />

The Rollup BSN contract is instantiated as follows:
```shell
INSTANT_MSG={
  "admin": <ADMIN_ADDRESS>,
  "bsn_id": <BSN_ID>,
  "max_msgs_per_interval": <MAX_MSGS_PER_INTERVAL>,
  "min_pub_rand": <MIN_PUB_RAND>,
  "rate_limiting_interval": <RATE_LIMITING_INTERVAL>,
  "allowed_finality_providers": [<FP_PUBKEY_1>, <FP_PUBKEY_2>] #Optional

> }

Reviewer 2: Don't think the above is valid shell code. This doesn't seem to be able to be copy-pasteable.

Author: Why?

Reviewer 2: If you paste this in a `.sh` file will this be valid code?

babylond tx wasm instantiate $CODE_ID '$INSTANT_MSG' \
  --from $ADMIN_KEY \
  --chain-id $BBN_CHAIN_ID \
  --label $LABEL \
  --admin $ADMIN_ADDRESS \

>   --fees 100000ubbn \

Reviewer 2: This is also a parameter the user can choose.

>   -y

Reviewer 2: We should not provide the prompt bypass flag here and allow the safety prompts to appear for users just copy pasting.

```

> The parameters:

Reviewer 2: These are the parameters of the instantiate message. Should we include a bullet point noting that the remaining parameters are standard transaction execution parameters, or perhaps list them as separate bullet points for clarity?

* `admin`: The Babylon Genesis address of the contract administrator
* `bsn_id`: The unique identifier for the BSN rollup
  Ensure that this is unique and not already used by another BSN on Babylon
  Genesis as this will affect your registration
* `min_pub_rand`: Minimum public randomness values included per public
  randomness commit by finality providers
* `rate_limiting_interval`: Length (in blocks) of the interval for
  rate-limiting submissions
* `max_msgs_per_interval`: Maximum number of submissions allowed per finality
  provider within each interval
* `allowed_finality_providers`: List of finality providers from which the contract will
  accept finality signatures and randomness

<!-- TODO: once we have docs for the allow list and rate limiting, we can link
those here -->

> 

Reviewer 2: ```suggestion
> ⚡ **Governance note**: In the case of a permissioned CosmWasm Babylon Genesis network,
> the contract instantiation should happen either via governance or via a governance allow-listed
> Babylon Genesis address.

```

## Rollup BSN Registration
<img width="3032" height="287" alt="register" src="https://github.com/user-attachments/assets/935fb9bd-f42d-47ba-a0f3-0fc628961e67" />

After deploying and instantiating the Rollup BSN contract, the
rollup must register on Babylon Genesis. Registration requires 
submitting metadata to identify and describe the BSN

To register the BSN, use:
```shell
babylond tx btcstkconsumer register-consumer \
                                <consumer_id> \
                                <consumer_name> \
                                <consumer_ description> \
                                <max_multi_staked_fps> \
                                <rollup_finality_contract_address>
```

Required metadata for BSN registration:

> * `consumer_id`: BSN ID of your rollup (same as the one you used on contract

Reviewer 2: ```suggestion
* `consumer_id`: BSN ID of your rollup (same as the BSN ID you used on contract
```

  instantiation)
* `consumer_name`: Human-readable name of your rollup BSN (e.g., `"DeFi Rollup Chain"`)
* `consumer_description`: Brief description of the rollup BSN's purpose
* `max_multi_staked_fps`: Maximum number of rollup FPs per BTC delegation. Prevents delegators from splitting
 their BTC across too many networks (recommended: `3-5`)

> <!-- TODO: this should not be here, we will have a global limit on fps -->

Reviewer 2: Let's just remove it?

> * `rollup_finality_contract_address`: Babylon Genesis address of the deployed Rollup BSN contract (`bbn1...` format)

Reviewer 2: ```suggestion
* `rollup_finality_contract_address`: Babylon Genesis address of the deployed
  Rollup BSN contract (`bbn1...` format)
```

> > **Note**: The above operation requires governance in permissioned
> > registration networks.

Reviewer 2: ```suggestion
> ⚡ **Governance Note**: The above operation can only be executed by governance
> in permissioned registration networks.
```

## Contract Maintenance
<img width="1848" height="677" alt="adminRole" src="https://github.com/user-attachments/assets/9b942068-3d72-4ae8-a47e-79894427204e" />

The Rollup BSN contract exposes `admin-only` functions for ongoing management. These allow you to clean up

>  old data, update roles, and adjust configuration without redeploying the contract.

Reviewer 2: ```suggestion
old data, update roles, and adjust configuration without redeploying the contract.
```

Each operation is executed with:

```shell

> babylond tx wasm execute <CONTRACT_ADDRESS> '<MSG>' \

Reviewer 2: We should say what `CONTRACT_ADDRESS` and `MSG` are.

  --from $ADMIN_KEY \
  --chain-id $BBN_CHAIN_ID \
  --keyring-backend test \
  --gas auto \
  --gas-adjustment 1.3 \
  --fees 1000000ubbn \

>   -y

Reviewer 2: These parameteres need to be cleaned up, specifically:
* --gas auto + --gas-adjustment + --fees do not make sense to go together.
* `-y` avoids a safety check
* `--keyring-backend test` assumes that the keyring is a test keyring.

I suggest we either don't provide those parameters at all (i.e., we just list `babylond tx wasm execute <CONTRACT_ADDRESS> '<MSG>'`) or we explain them. Given that this is not a guide on how to send cosmos txs, I suggest we do the former and keep things simple.

Same for all other commands.

```

### Modifying the Contract Administrator
```shell
MSG={
  "update_admin": {
    "admin": <NEW_ADMIN_ADDRESS>
  }
}
```
Transfers contract admin rights to a new Babylon address. Typically used when rotating keys or migrating

>  ownership to another entity

Reviewer 2: ```suggestion
ownership to another entity
```

there are other places with preceding spaces

### Data Pruning
```shell
MSG={
  "prune_data": {
    "rollup_height": <BLOCK_HEIGHT>,
    "max_signatures_to_prune": <MAX_SIGS>, # optional
    "max_pub_rand_values_to_prune": <MAX_PUB_RAND> # optional
  }
}
```
Removes all finality signatures and public randomness for rollup blocks with height less than or

>  equal to the specified value. Useful for reducing on-chain storage as the chain grows. Optional
>   fields control pruning batch size to avoid gas exhaustion.

Reviewer 2: ```suggestion
equal to the specified value. Useful for reducing on-chain storage as the chain grows. Optional
fields control pruning batch size to avoid gas exhaustion.
```

See [PRUNING.m](./PRUNING.md) for detailed guidelines on selecting safe pruning heights and best

>  practices

Reviewer 2: ```suggestion
practices.
```

Missing periods in ends of sentences

### Modifying the Finality Providers Allow-List
```shell
MSG={
  "add_to_allowlist": {
    "fp_pubkey_hex_list": [
      <FP_PUBKEY>
    ]
  }
}

> ```
> ```shell
> MSG={

Reviewer 2: ```suggestion
MSG={
```

  "remove_from_allowlist": {
    "fp_pubkey_hex_list": [
      <FP_PUBKEY>
    ]
  }
}
```
Adds or removes Finality Providers from the allowlist. Use this to onboard new providers or revoke

>  access from inactive or misbehaving ones

Reviewer 2: ```suggestion
access from inactive or misbehaving ones
```

> ### Admin Queries

Reviewer 2: These are not admin queries, but general queries. We don't really need this section here

In addition to executing messages, admins can query contract state to inspect configuration and

>  permissions. Use:

Reviewer 2: ```suggestion
permissions. Use:
```

```shell
babylond query wasm contract-state smart <CONTRACT_ADDRESS> '<QUERY_MSG>' \
  --node $NODE_RPC \
  --output json
```
**Available admin queries:**
```shell
QUERY_MSG={
  "admin": {}
}
```
Returns the current contract admin address

```shell
QUERY_MSG={
  "config": {}
}
```
Returns the contract configuration, including BSN ID, rate limits, and other parameters

```shell
QUERY_MSG={
  "allowed_finality_providers": {}
}
```
Returns a list of BTC public keys (in hex format) currently allowlisted to submit data

> ## External Integration

Reviewer 2: I'm not really sure as to why we need this section. Even if we do, it surely does not need to be so technical. This is an operator's document on how to maintain a rollup BSN contract. The technical details of the messages that finality providers or the queries that the gadgets are doing are not really relevant.

Author: Let's remove it. But we should have dedicated doc explaining not how instantiate and deploy but also how to interact. What queries exists and what messages you can send

<img width="2748" height="529" alt="external integrations" src="https://github.com/user-attachments/assets/2904a675-f980-47ff-9c62-2bfab5fcf1a8" />

The Rollup BSN contract is used by two main types of off-chain actors:

### Finality Providers
Finality Providers are off-chain programs that watch rollup blocks, sign them, and submit data

>  on-chain. They must be registered and allowlisted to submit randomness and signatures

Reviewer 2: ```suggestion
on-chain. They must be registered and allowlisted to submit randomness and signatures
```

They interact with the contract by sending the following messages:
```shell
MSG={
  "commit_public_randomness": {
    "commitment": <MERKLE_ROOT>,
    "fp_pubkey_hex": <FP_BTC_PUBKEY>,
    "num_pub_rand": <COUNT>,
    "signature": <SIGNATURE>,
    "start_height": <START_BLOCK>
  }
}
```
Commits a Merkle root of public randomness values that will later be revealed in signature

>  submissions. This prevents Finality Providers from generating randomness adaptively based on block contents.

Reviewer 2: ```suggestion
submissions. This prevents Finality Providers from generating randomness adaptively based on block contents.
```

```shell
MSG={
  "submit_finality_signature": {
    "block_hash": <BLOCK_HASH>,
    "fp_pubkey_hex": <FP_BTC_PUBKEY>,
    "height": <BLOCK_HEIGHT>,
    "proof": <MERKLE_PROOF>,
    "pub_rand": <PUBLIC_RANDOMNESS>,
    "signature": <SIGNATURE>,
    "l1_block_hash_hex": <L1_HASH>, # optional
    "l1_block_number": <L1_NUMBER>  # optional
  }
}
```
Submits a finality signature for a rollup block, proving that it has been verified using previously

>  committed randomness. Includes Merkle proof to bind the randomness to the earlier commitment

Reviewer 2: ```suggestion
committed randomness. Includes Merkle proof to bind the randomness to the earlier commitment
```

### Finality Gadget

Finality Gadget is an off-chain service that tracks Voting table and which Finality Providers have

>  signed each block. It queries the contract for public randomness and voting data, and determines if
>   a block is finalized when more than 2/3 of the registered voting power has signed it
> 

Reviewer 2: ```suggestion
signed each block. It queries the contract for public randomness and voting data, and determines if
a block is finalized when more than 2/3 of the registered voting power has signed it

```

To query the contract, use:
```shell
babylond query wasm contract-state smart <CONTRACT_ADDRESS> '<QUERY_MSG>' \
  --node $NODE_RPC \
  --output json
```

**Available query messages:**
```shell
QUERY_MSG={
  "block_voters": {
    "height": <BLOCK_HEIGHT>,
    "hash_hex": <BLOCK_HASH_HEX>
  }
}
```
Returns a list of Finality Providers that submitted signatures for the specified block

```shell
QUERY_MSG={
  "first_pub_rand_commit": {
    "btc_pk_hex": <FP_BTC_PUBKEY>
  }
}
```
Returns the first public randomness commitment (if any) for the given Finality Provider

```shell
QUERY_MSG={
  "last_pub_rand_commit": {
    "btc_pk_hex": <FP_BTC_PUBKEY>
  }
}
```
Returns the most recent public randomness commitment for the given Finality Provider

```shell
QUERY_MSG={
  "list_pub_rand_commit": {
    "btc_pk_hex": <FP_BTC_PUBKEY>,
    "start_after": <START_HEIGHT>, # optional
    "limit": <LIMIT>, # optional, default 10, max 30
    "reverse": <REVERSE_ORDER> # optional, default false
  }
}
```
Returns a paginated list of public randomness commitments. Useful for coordinating with BTC timestamping

>  and verifying available randomness for signature validation

Reviewer 2: ```suggestion
and verifying available randomness for signature validation
```
